actionpack/test/controller/parameters/accessors_test.rb

```diff
@@ -419,4 +418,28 @@ class ParametersAccessorsTest < ActiveSupport::TestCase
     @params.dig(:person, :addresses)[0] = { city: "Boston", state: "Massachusetts" }
     assert_equal "Boston", @params.dig(:person, :addresses, 0, :city)
   end
+
+  test "has_value? works with hash argument after conversion" do
+    without_deprecated_parameters_hash_equality do
+      params = ActionController::Parameters.new(foo: { bar: "baz" })
+      assert params.has_value?("bar" => "baz")
```

Comment:
If I understand correctly, this _doesn't_ raise a deprecation warning, but the next line does. I think both should. The deprecation behaviour shouldn't depend on the state of the params object.

Reply:
>  this doesn't raise a deprecation warning, but the next line does


It is correct while the deprecated code is enabled. But within the test (inside of without_deprecated_parameters_hash_equality) the deprecated behaviour is disabled so no warnings.

The goal of the change was to allow users to not care about internal conversions  _after_ the deprecated code is removed.
But your point makes sense. Before the deprecation is removed users are going to be affected by the warnings.

I reworked has_value? using `each_value` which converts all values to AC::Params. This way we should avoid both the warnings and comparison inconsistencies when the deprecated code is removed.

Wdyt about the performance impact of this new has_value? implementation.

Reply:
> Wdyt about the performance impact of this new has_value? implementation.

This makes more sense to me. The perf difference shouldn't matter much because `has_value?` isn't a common operation on parameters.

I think we can also cut this down to 2 tests:

```
  test "has_value? converts hashes to parameters" do
    with_parameters_hash_equality_deprecation do
      assert_not_deprecated do
        params = ActionController::Parameters.new(foo: { bar: "baz" })
        assert params.has_value?("bar" => "baz")
        params[:foo] # converts value to AC::Params
        assert params.has_value?("bar" => "baz")
      end
    end
  end

  test "has_value? works with parameters" do
    with_parameters_hash_equality_deprecation do
      assert_not_deprecated do
        params = ActionController::Parameters.new(foo: { bar: "baz" })
        assert params.has_value?(ActionController::Parameters.new("bar" => "baz"))
      end
    end
  end
```

Reply:
within the `with_parameters_hash_equality_deprecation` opt-in the `assert_not_deprecated` cannot raise. It doesn't make sense to combine them. I think we could cut to those 2, wdyt?

```
  test "has_value? converts hashes to parameters" do
    assert_not_deprecated do
      params = ActionController::Parameters.new(foo: { bar: "baz" })
      assert params.has_value?("bar" => "baz")
      params[:foo] # converts value to AC::Params
      assert params.has_value?("bar" => "baz")
    end
  end

  test "has_value? works with parameters" do
    with_parameters_hash_equality_deprecation do
      params = ActionController::Parameters.new(foo: { bar: "baz" })
      assert params.has_value?(ActionController::Parameters.new("bar" => "baz"))
    end
  end

```

Reply:
> within the with_parameters_hash_equality_deprecation opt-in the assert_not_deprecated cannot raise. It doesn't make sense to combine them.

🤦  Sorry, I misunderstood the behaviour of the config. Let's revert the name change on `with_parameters_hash_equality_deprecation` to something like `without_deprecated_parmas_hash_equality` and use the two tests you mentioned. Thanks!

Reply:
Apologies about the confusing naming of the flag.

I reverted the method name change and added a changelog.